### PR TITLE
ICS 101: interchain Liquidity

### PR DESCRIPTION
**Pull request by liangping**
ICS-101 Interchain Swaps enables chains their own token pricing mechanism and exchange protocol via IBC transactions. Each chain can thus play a role in a fully decentralised exchange network.

### CODE REVIEW

**Comment by crodriguezvega** [comment]
Thank you for submitting this new spec, @liangping. I will try to give it a first review this quarter, but realistically you should expect that it gets merged somewhere in Q3.

**Comment by liangping** [comment]
Got it. Thanks

**Comment by liangping** [comment]
> too low level about implementation about pools to tbh

any tips for improvement?

**Comment by dzmitry-lahoda** [comment]
you specifying that there is pool.with xyk. why  not making swap to be buy or sell messages. we do  not care what inside. as long there limits satisifed, protocol is correct. given any packet can be async one, now ack on same block. you have timeout already too for orders.

**Comment by dzmitry-lahoda** [comment]
just may be remove  on essential text our of spec about xyc

**Comment by liangping** [comment]
> you specifying that there is pool.with xyk. why not making swap to be buy or sell messages. we do not care what inside. as long limits satisfied, protocol is correct. given any packet can be async one, no ack on same block is fact. and you have timeout already too for orders.

I'm not entirely sure I understand, but we have ICS 100 which enables the sending of buy/sell messages to the counterparty chain.

This specification is designed to facilitate quick asset swaps between two chains without requiring voucher assets. Currently, we do not have plans to implement complex pools like those found in DEFI Hubs such as Osmosis or Uniswap.

**Comment by dzmitry-lahoda** [comment]
> > you specifying that there is pool.with xyk. why not making swap to be buy or sell messages. we do not care what inside. as long limits satisfied, protocol is correct. given any packet can be async one, no ack on same block is fact. and you have timeout already too for orders.
> 
> I'm not entirely sure I understand, but we have ICS 100 which enables the sending of buy/sell messages to the counterparty chain.
> 
> This specification is designed to facilitate quick asset swaps between two chains without requiring voucher assets. Currently, we do not have plans to implement complex pools like those found in DEFI Hubs such as Osmosis or Uniswap.
> 
> But we are happy if we can work together to improve this spec.

i do not refer to ics 100, nor vouchers, nor telling of any protocol requiring 2 ibc packets.

why you are implemwnting pools inside ibc protocol?

**Comment by DedicatedDev** [comment]
> > > you specifying that there is pool.with xyk. why not making swap to be buy or sell messages. we do not care what inside. as long limits satisfied, protocol is correct. given any packet can be async one, no ack on same block is fact. and you have timeout already too for orders.
> > 
> > 
> > I'm not entirely sure I understand, but we have ICS 100 which enables the sending of buy/sell messages to the counterparty chain.
> > This specification is designed to facilitate quick asset swaps between two chains without requiring voucher assets. Currently, we do not have plans to implement complex pools like those found in DEFI Hubs such as Osmosis or Uniswap.
> > But we are happy if we can work together to improve this spec.
> 
> i do not refer to ics 100, nor vouchers, nor telling of any protocol requiring 2 ibc packets.
> 
> why you are implemwnting pools inside ibc protocol?



> you specifying that there is pool.with xyk. why not making swap to be buy or sell messages. we do not care what inside. as long limits satisfied, protocol is correct. given any packet can be async one, no ack on same block is fact. and you have timeout already too for orders.

The use of "pool.with xyk" in the specification is to indicate the involvement of the XYK (Automated Market Maker) pool in the swapping process. The specification may use "pool.with xyk" to abstract the underlying implementation details of the buy and sell operations.

By using "pool.with xyk" instead of explicit "buy" or "sell" messages, the specification remains agnostic about the specific actions being taken. This allows for flexibility in the implementation and supports different types of swapping mechanisms.

The focus of the specification is on ensuring that the necessary limits and protocols are satisfied for the swapping process. As long as the requirements are met and the protocol is correctly followed, the specific actions being performed (buying or selling) can be abstracted within the "pool.with xyk" implementation.

Regarding the suggestion to remove non-essential text from the specification about "xyk," it can be considered if it does not impact the clarity and completeness of the specification. However, it is important to ensure that the essential information about the involvement of the XYK pool and its purpose is adequately conveyed in the specification.

In summary, the use of "pool.with xyk" in the specification allows for flexibility in the implementation of the swapping process while ensuring that the necessary limits and protocols are satisfied. The decision to abstract the actions as "pool.with xyk" instead of explicit "buy" or "sell" messages enables different swapping mechanisms and simplifies the specification by focusing on the essential requirements.


Note: ics-100 is just order book, ics-101 is just AMM. 
I'm not sure how to answer about your question.
"There are order book, why do we have to use AMM" ????

**Comment by DedicatedDev** [comment]
> too low level about implementation about pools to tbh

It's just initial steps to reflects weighted pools, there are continuously improvement to support various swap algorithm.

**Comment by dzmitry-lahoda** [comment]
https://github.com/osmosis-labs/osmosis-rust/blob/f686934e74a7abc5695775e8995e6d4412086f9b/packages/osmosis-std/src/types/osmosis/gamm/v1beta1.rs#L410 

https://github.com/osmosis-labs/osmosis-rust/blob/f686934e74a7abc5695775e8995e6d4412086f9b/packages/osmosis-std/src/types/osmosis/gamm/v1beta1.rs#L448

see buy and sell on stable swap and constant swap

we were thinking about xc exchanges a lot - seems like pool/amm specific for swap is not so useful

other ecosystem also does not tells if that pool or not
https://github.com/paritytech/xcm-format/blob/master/README.md#exchangeasset

**Comment by DedicatedDev** [comment]
> https://github.com/osmosis-labs/osmosis-rust/blob/f686934e74a7abc5695775e8995e6d4412086f9b/packages/osmosis-std/src/types/osmosis/gamm/v1beta1.rs#L410
> 
> https://github.com/osmosis-labs/osmosis-rust/blob/f686934e74a7abc5695775e8995e6d4412086f9b/packages/osmosis-std/src/types/osmosis/gamm/v1beta1.rs#L448
> 
> see buy and sell on stable swap and constant swap
> 
> same here https://github.com/apollodao/cw-dex/
> 
> we were thinking about xc exchanges a lot - seems like pool/amm specific for swap is not so useful
> 
> other ecosystem also does not tells if that pool or not https://github.com/paritytech/xcm-format/blob/master/README.md#exchangeasset
> 
> if xyk/amm or anything else does swap is not essential
> 
> if you want to specify atomicity, you can still have buy/sell but name protoocl atomic swap. but the problem, you cannot be atomic as target module may be async, and do not confirm nor timeout packet, so one may not have that too. really it is not atomic but single packet buy or sell

To address the feedback and establish the significance of your work in creating a generalized crosschain AMM package as a standard for ICS-101 (IBC-Go), you can provide the following response:

Thank you for sharing the links and providing feedback on the existing implementations of AMMs and cross-chain exchanges. The references you shared from the Osmosis Rust repository demonstrate specific implementations of AMMs using buy and sell operations for stable swap and constant swap mechanisms. These implementations are valuable examples that showcase the functionality and behavior of AMMs in those specific contexts.

However, the goal of creating a generalized crosschain AMM package as a standard for ICS-101 (IBC-Go) is to provide a broader solution that can be utilized in various AMM implementations across different ecosystems. By developing a standardized approach, you aim to enhance interoperability and facilitate seamless asset exchange across different blockchain networks.

While specific AMM implementations may focus on the details of individual pools or AMM types, the emphasis of your work is on providing a generalized framework that can be adopted by different projects and ecosystems. This framework will allow for cross-chain asset swaps without being tied to a particular AMM pool or implementation. It provides a common interface and standard protocols for interacting with AMMs, enabling easy integration and compatibility across diverse blockchain networks.

The mentioned XCM Format from the Parity Technologies repository is another valuable resource in the realm of cross-chain communication and asset exchange. It further reinforces the importance of establishing interoperable standards to enable efficient cross-chain operations.

By creating a generalized crosschain AMM package as a standard for ICS-101 (IBC-Go), you contribute to the broader blockchain ecosystem by providing a consistent and interoperable solution for asset exchange. Your work aims to streamline the development and adoption of AMM functionality across different projects and networks, facilitating increased liquidity and cross-chain functionality.

In summary, while existing AMM implementations focus on specific pools and mechanisms, your work on a generalized crosschain AMM package as a standard for ICS-101 (IBC-Go) serves a broader purpose by establishing interoperable protocols and interfaces. This standardized approach enhances cross-chain asset exchange and contributes to the overall advancement of the blockchain ecosystem

**Comment by dzmitry-lahoda** [comment]
yes, thank you.  just to be sure this ics supports(does not prevents) any kind of swaps and hoppefully fully compatible with ics20 assets(unlike ics100).

File: spec/app/ics-101-interchain-swap/README.md [hunk 1]
```diff
@@ -0,0 +1,1613 @@
...
+    pool :InterchainLiquidityPool
+    static initialize(pool: InterchainLiquidityPool) : InterchainMarketMaker {
+        return {
+            pool: pool
+        }
+    }
+
+    // MarketPrice Bi / Wi / (Bo / Wo)
+    function marketPrice(denomIn, denomOut string): float64 {
+        const tokenIn = this.Pool.findAssetByDenom(denomIn)
```
Comment (crodriguezvega) [inline spec/app/ics-101-interchain-swap/README.md:285]: I think the `findAssetByDenom` function definition is missing from `InterchainLiquidityPool` interface.

File: spec/app/ics-101-interchain-swap/README.md [hunk 2]
```diff
@@ -0,0 +1,1613 @@
...
+        const balanceOut = tokenOut.balance.amount
+        const weightIn := tokenIn.weight
+        const weightOut := tokenOut.weight
+
+        return balanceIn / weightIn / (balanceOut / weightOut)
+    }
+
+    // P_issued = P_supply * ((1 + At/Bt) ** Wt -1)
+    function depositSingleAsset(token: Coin): Coin {
+        const asset = this.pool.findAssetByDenom(token.denom)
```
Comment (crodriguezvega) [inline spec/app/ics-101-interchain-swap/README.md:297]: I guess the function should return null if `findAssetByDenom` fails to find the denomination, right?

File: spec/app/ics-101-interchain-swap/README.md [hunk 3]
```diff
@@ -0,0 +1,1613 @@
...
+          const msgTakeMultiDeposit: MsgTakeMultiAssetDepositRequest = protobuf.decode(data.Data);
+          const resTakeMultiDeposit: MsgTakePoolResponse = protobuf.decode(ack.GetResult());
+          const errTakeMultiDeposit = store.OnTakeMultiAssetDepositAcknowledged(msgTakeMultiDeposit, data.StateChange);
+          abortTransactionUnless(errTakeMultiDeposit === undefined);
+          break;
+
+        case "MULTI_WITHDRAW":
+          const msgMultiWithdraw: MsgMultiAssetWithdrawRequest = protobuf.decode(data.Data);
+          const resMultiWithdraw: MsgMultiAssetWithdrawResponse = protobuf.decode(ack.GetResult());
+          const errMultiWithdraw = store.OnMultiAssetWithdrawAcknowledged(msgMultiWithdraw, resMultiWithdraw);
```
Comment (crodriguezvega) [inline spec/app/ics-101-interchain-swap/README.md:682]: I think we're missing a definition for `OnMultiAssetWithdrawAcknowledged` function.

File: spec/app/ics-101-interchain-swap/README.md [hunk 4]
```diff
@@ -0,0 +1,1613 @@
...
+          abortTransactionUnless(errMultiWithdraw === undefined);
+          break;
+
+        case "LEFT_SWAP":
+        case "RIGHT_SWAP":
+          const msgSwap: MsgSwapRequest = protobuf.decode(data.Data);
+          const resSwap: MsgSwapResponse = protobuf.decode(ack.GetResult());
+          const errSwap = store.OnSwapAcknowledged(msgSwap, resSwap);
```
Comment (crodriguezvega) [inline spec/app/ics-101-interchain-swap/README.md:690]: I think definition of `OnSwapAcknowledged` is also missing.

File: spec/app/ics-101-interchain-swap/README.md [hunk 5]
```diff
@@ -0,0 +1,1613 @@
...
+
+  pool.subtractAsset(stateChange.out[0]);
+  pool.addAsset(msg.tokenIn);
+
+  store.setInterchainLiquidityPool(pool);
+
+  return { tokens: stateChange.out };
+}
+
+function onMakePoolAcknowledged(msg: MsgMakePoolRequest, poolId: string): void {
```
Comment (crodriguezvega) [inline spec/app/ics-101-interchain-swap/README.md:1407]: If the make pool flow completes successfully initiated from chain A, but there is not take pool executed from the counterparty chain B. Is there a way for the creator of the pool on chain A, to recover their tokens?

File: spec/app/ics-101-interchain-swap/README.md [hunk 6]
```diff
@@ -0,0 +1,1613 @@
...
+  decimal: int32;
+}
+```
+
+```ts
+interface InterchainLiquidityPool {
+  id: string;
+  sourceCreator: string;
+  destinationCreator: string;
+  assets: []PoolAsset;
```
Comment (crodriguezvega) [inline spec/app/ics-101-interchain-swap/README.md:155]: This array should always have 2 elements, right? Would it be better for readability and ease of comprehension to actually make two different fields (e.g `sourcePoolAsset` and `destinationPoolAsset`)? Just to avoid working with all these indexes...

**Review by crodriguezvega — COMMENTED** [review]
Sorry for taking so long to give a review and thank you for your patience, @liangping. I just completed the first review (although I couldn't finish it all). This is a pretty complex protocol, so I will need a few more passes. For now I have left a few suggestions and questions.

**Comment by liangping** [comment]
thank you. will improve it soon.

File: spec/app/ics-101-interchain-swap/README.md [hunk 7]
```diff
@@ -0,0 +1,1613 @@
...
+
+![Interchain Swap Diagram](interchain-swap.svg)
+
+### Algorithms
+
+#### Invariant
+
+A constant invariant is maintained after trades, taking into consideration token weights and balance. The value function $V$ is defined as:
+
+$$V = {&Pi;_tB_t^{W_t}}$$
```
[the remark this answers is not in the document]
Comment (liangping) [reply spec/app/ics-101-interchain-swap/README.md:75]: $$V = {&Pi;_tB_t^{W_t}}$$ is actually render like `V` = `(B0 ** W0) * (B1 ** W1) ...`

File: spec/app/ics-101-interchain-swap/README.md [hunk 8]
```diff
@@ -0,0 +1,1613 @@
...
+
+These are methods that handle packets relayed from a source chain, and includes pool state updates and token transfers. In this way, packets relayed on the source chain update pool states on the destination chain.
+
+```ts
+function onMakePoolReceived(msg: MsgMakePoolRequest, poolID: string, sourceChainId: string): string {
+  abortTransactionUnless(msg.validateBasic() === undefined);
+  const { pool, found } = store.getInterchainLiquidityPool(poolID);
+  abortTransactionUnless(msg.validateBasic() === undefined);
+
+  const liquidityBalance = msg.liquidity[1].balance;
```
[the remark this answers is not in the document]
Comment (DedicatedDev) [reply spec/app/ics-101-interchain-swap/README.md:1243]: As we have plans to support multiple assets in the future, hardcoding each token as a field in the message structure is not a scalable or reasonable approach. Instead, we will adopt a 'promised order' method to streamline this process.

In this method, '0' will always represent the source chain, and '1' will always represent the target chain. This way, we can dynamically support various tokens without needing to alter the message structure, making the system more flexible and easier to maintain as we add more assets.

By adopting this approach, we aim to create a more adaptable and scalable system that can readily accommodate multiple assets without the need for significant structural changes.

[the remark this answers is not in the document]
Comment (DedicatedDev) [reply spec/app/ics-101-interchain-swap/README.md:1243]: later we can divide source chain assets and destination assets, but in this simple version, 0 imply source chain 1 imply target chain

File: spec/app/ics-101-interchain-swap/README.md [hunk 9]
```diff
@@ -0,0 +1,1613 @@
...
+}
+}
+```
+
+```ts
+function generatePoolId(sourceChainId: string, destinationChainId: string, denoms: string[]): string {
+  const chainPrefix = [sourceChainId, destinationChainId].sort().join(",");
+  const id = chainPrefx.concat(denoms.sort().join(","));
+  const poolId = "pool" + sha256.hash(id);
+  return poolId;
```
[the remark this answers is not in the document]
Comment (DedicatedDev) [reply spec/app/ics-101-interchain-swap/README.md:266]: We have opted not to hardcode paths, unlike some other solutions. The reason for this choice is to allow for the possibility of using various channels, which can be selected by the relayer on the frontend side. By doing so, we aim to provide greater flexibility to the frontend, thereby enhancing the robustness and reliability of the entire system, including the frontend components.

File: spec/app/ics-101-interchain-swap/README.md [hunk 10]
```diff
@@ -0,0 +1,1613 @@
...
+
+    abortTransactionUnless(sourceLiquidity.amount > msg.liquidity[0].balance.amount)
+
+
+    const lockErr = store.lockTokens(msg.sourcePort, msg.sourceChannel, senderAddress, msg.liquidity[0].balance);
+
+    abortTransactionUnless(lockErr === undefined)
+
+    const packet: IBCSwapPacketData = {
+      type: "MAKE_POOL",
```
[the remark this answers is not in the document]
Comment (DedicatedDev) [reply spec/app/ics-101-interchain-swap/README.md:953]: yes, but current added as a string

File: spec/app/ics-101-interchain-swap/README.md (hunk 2 above)
Comment (DedicatedDev) [reply spec/app/ics-101-interchain-swap/README.md:297]: it is helper function, it doesn't change  pool status directly. so it has to return amount. 
based on this calculate, it will change in message part.

File: spec/app/ics-101-interchain-swap/README.md (hunk 5 above)
Comment (DedicatedDev) [reply spec/app/ics-101-interchain-swap/README.md:1407]: yes we've adding cancel logic.

File: spec/app/ics-101-interchain-swap/README.md (hunk 6 above)
Comment (DedicatedDev) [reply spec/app/ics-101-interchain-swap/README.md:155]: balancer is for multi-asset deposit. it doesn't mean only 2 tokens. 
of course, currently we only support 2 tokens. 
but in future, we will support several tokens. 
I agree define sourcePoolAsset, or destinationPooAsset, but it will become array.

File: spec/app/ics-101-interchain-swap/README.md [hunk 11]
```diff
@@ -0,0 +1,1635 @@
...
+  // Validate message
+  const validationErr = msg.validateBasic();
+  abortTransactionUnless(validationErr === undefined);
+
+  // Check if pool token denom exists
+  const poolTokenDenom = msg.poolToken.denom;
+  abortTransactionUnless(store.bankKeeper.hasSupply(poolTokenDenom));
+
+  // Get the liquidity pool
+  const { pool, found } = k.getInterchainLiquidityPool(ctx, poolTokenDenom);
```
Comment (crodriguezvega) [inline spec/app/ics-101-interchain-swap/README.md:1168]:
```suggestion
  const { pool, found } = store.getInterchainLiquidityPool(msg.poolId)
```

File: spec/app/ics-101-interchain-swap/README.md [hunk 12]
```diff
@@ -0,0 +1,1635 @@
...
+}
+
+
+function swap(msg: MsgSwapRequest): MsgSwapResponse {
+  const { pool, found } = store.getInterchainLiquidityPool(msg.poolId);
+  abortTransactionUnless(found);
+
+  abortTransactionUnless(pool.status === "ACTIVE");
+
+  const lockErr = store.lockTokens(pool.counterPartyPort, pool.counterPartyChannel, msg.sender, msg.tokenIn);
```
Comment (crodriguezvega) [inline spec/app/ics-101-interchain-swap/README.md:1198]: Is this correct that tokens are locked when the user does both a left or a right swap? 

If I understand it correctly, in the case of a right swap, the user should specify how many tokens they want to buy, and the user doesn't know yet how much it needs to pay for them, right?

**Review by crodriguezvega — COMMENTED** [review]
Spent again a bit of time on the review. Still need to spend more time in the coming weeks, but I left for now a few more comments. :)

File: spec/app/ics-101-interchain-swap/README.md (hunk 12 above)
Comment (DedicatedDev) [reply spec/app/ics-101-interchain-swap/README.md]: They can know this. Market function include swap fee and then they can estimate output amount. 
In go implementation, Market functions is just stateless(don't update state variable directly and provide only estimation value). so before sending swap request, they can estimate how much they can get. 
Based on that, they can make swap request. 

Also based on slippage, they can avoid to fail in swap.

File: spec/app/ics-101-interchain-swap/README.md [hunk 13]
```diff
@@ -0,0 +1,1635 @@
...
+$$SP_i^o = (B_i/W_i)/(B_o/W_o)$$
+
+- $B_i$ is the balance of token $i$, the token being sold by the trader which is going into the pool
+- $B_o$ is the balance of token $o$, the token being bought by the trader which is going out of the pool
+- $W_i$ is the weight of token $i$
+- $W_o$ is the weight of token $o$
+
+#### Fees
+
+Traders pay swap fees when they trade with a pool. These fees can be customized with a minimum value of 0.01% and a maximum value of 10%. Basis points for `feeRate` is used, where the integer 1 is 0.01% and integer 100 is 1.00%.
```
[the remark this answers is not in the document]
Comment (liangping) [reply spec/app/ics-101-interchain-swap/README.md]: Yeah. 1000 is the maximum value for feeRate

File: spec/app/ics-101-interchain-swap/README.md [hunk 14]
```diff
@@ -0,0 +1,1635 @@
...
+`onRecvPacket` is called by the routing module when a packet addressed to this module has been received.
+
+```ts
+function OnRecvPacket(packet: Packet): Uint8Array | undefined {
+  const data: IBCSwapPacketData = packet.data;
+  switch (data.type) {
+    case "MAKE_POOL":
+      const makePoolMsg: MsgMakePoolRequest = protobuf.decode(MsgMakePoolRequest, data.Data);
+      abortTransactionUnless(data.stateChange.poolId === "");
+      abortTransactionUnless(store.has(data.stateChange.poolId)); // existed already.
```
[the remark this answers is not in the document]
Comment (liangping) [reply spec/app/ics-101-interchain-swap/README.md]: Yeah. it's `abortTransactionUnless(data.stateChange.poolId !== ""`

File: spec/app/ics-101-interchain-swap/README.md [hunk 15]
```diff
@@ -0,0 +1,1635 @@
...
+}
+```
+
+These are methods that handle packets relayed from a source chain, and includes pool state updates and token transfers. In this way, packets relayed on the source chain update pool states on the destination chain.
+
+```ts
+function onMakePoolReceived(msg: MsgMakePoolRequest, poolID: string, sourceChainId: string): string {
+  abortTransactionUnless(msg.validateBasic() === undefined);
+  const { pool, found } = store.getInterchainLiquidityPool(poolID);
+  abortTransactionUnless(msg.validateBasic() === undefined);
```
[the remark this answers is not in the document]
Comment (liangping) [reply spec/app/ics-101-interchain-swap/README.md]: yes.

File: spec/app/ics-101-interchain-swap/README.md [hunk 16]
```diff
@@ -0,0 +1,1635 @@
...
+  abortTransactionUnless(found);
+
+  pool.status = "ACTIVE";
+  const asset = pool.findPoolAssetBySide("DESTINATION");
+  abortTransactionUnless(asset === undefined);
+
+  const totalAmount = pool.sumOfPoolAssets();
+  const mintAmount = (totalAmount * asset.weight) / 100;
+
+  store.mintTokens(pool.sourceCreator, new sdk.Coin(pool.supply.denom, mintAmount));
```
[the remark this answers is not in the document]
Comment (liangping) [reply spec/app/ics-101-interchain-swap/README.md]: I am considering whether we should include an option for depositors to select where their LP tokens are minted.

File: spec/app/ics-101-interchain-swap/README.md [hunk 17]
```diff
@@ -0,0 +1,1635 @@
...
+
+    const channelValidationErr = host.ChannelIdentifierValidator(msg.SourceChannel);
+
+    abortTransactionUnless(channelValidationErr === undefined)
+
+    const validationErr = msg.ValidateBasic();
+
+    abortTransactionUnless(validationErr === undefined)
+
+    abortTransactionUnless(store.hasSupply(msg.liquidity[0].balance.denom))
```
[the remark this answers is not in the document]
Comment (liangping) [reply spec/app/ics-101-interchain-swap/README.md]: updated

File: spec/app/ics-101-interchain-swap/README.md [hunk 18]
```diff
@@ -0,0 +1,1635 @@
...
+
+
+    const sourceLiquidity = store.GetBalance(msg.creator, msg.liquidity[0].balance.denom);
```
[the remark this answers is not in the document]
Comment (liangping) [reply spec/app/ics-101-interchain-swap/README.md]: updated

File: spec/app/ics-101-interchain-swap/README.md [hunk 19]
```diff
@@ -0,0 +1,1635 @@
...
+  return { tokens: stateChange.out };
+}
+
+function onMakePoolAcknowledged(msg: MsgMakePoolRequest, poolId: string): void {
+  const pool = new InterchainLiquidityPool(
+    poolId,
+    msg.creator,
+    msg.counterPartyCreator,
+    store,
+    msg.liquidity,
```
[the remark this answers is not in the document]
Comment (liangping) [reply spec/app/ics-101-interchain-swap/README.md]: updated

File: spec/app/ics-101-interchain-swap/README.md [hunk 20]
```diff
@@ -0,0 +1,1635 @@
...
+    msg.swapFee,
+    msg.sourcePort,
+    msg.sourceChannel
+  );
+
+  pool.sourceChainId = store.chainID();
+
+  const totalAmount = sdk.NewInt(0);
+  for (const asset of msg.liquidity) {
```
[the remark this answers is not in the document]
Comment (liangping) [reply spec/app/ics-101-interchain-swap/README.md]: updated

File: spec/app/ics-101-interchain-swap/README.md (hunk 11 above)
Comment (liangping) [reply spec/app/ics-101-interchain-swap/README.md]: updated
